crates/sui-core/src/authority/authority_store.rs

```diff
@@ -1152,15 +1217,49 @@ impl<S: Eq + Serialize + for<'de> Deserialize<'de>> SuiDataStore<S> {
 
         // Schedule the certificate for execution
         self.add_pending_certificates(vec![(transaction_digest, Some(certificate.clone()))])?;
+
+        // Holding _tx_lock avoids the following race:
+        // - we check effects_exist, returns false
+        // - another task (starting from handle_node_sync_certificate) writes effects,
+        //    and then deletes locks from assigned_object_versions
+        // - we write to assigned_object versions, re-creating the locks that were just deleted
+        // - now it's possible to run a new tx against old versions of the shared objects.
+        let _tx_lock = self.acquire_tx_lock(&transaction_digest).await;
```

Comment:
Do we not need to explicitly release this lock?

Reply:
I thought the implicit drop would add the tx to the retry queue

Reply:
Ah, no, this is a new method I added `acquire_tx_lock` which gets the lock only. `begin_tx` holds the lock and writes to the WAL